tools/hlint.sh

```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env bash
+
+usage() { echo "Usage: $0 -f [all, changeset] -m [check, inplace]" 1>&2; exit 1; }
+
+files=''
+check=true
+
+while getopts ':f:m:' opt
+ do
+     case $opt in
+         f) f=${OPTARG}
+            if [ "$f" = "all" ]; then
+              files=$(find libs/ services/ -not -path "*/test/*" -name "*.hs")
+            elif [ "$f" = "changeset" ]; then
+              files=$(git diff --name-only HEAD | grep \.hs\$)
```

Comment:
Running `-f all` doesn't lint tests, but running `-f changeset` runs tests. Is this correct?
Aside from that, why not lint tests?

Reply:
> Aside from that, why not lint tests?

I'd very much like to have tests linted, but separate PR maybe?  Get it to work first?

Reply:
The reason behind ignoring tests is that we currently have certain amount of "unsafe code" that's perfectly safe inside their test environments. Most notably we use index accessors a lot when dealing with prekeys in tests. 

Should I add back test folders to the all flag or remove it from changeset?

I left it like this bc when going over the whole Codebase I didn't want to have to rework all those test cases or add a bunch of ignore rules. But I did want my own tests to be linted. 

I should've brought that up in the chat :)

Reply:
as i said, this can be done separately.

the question that remains is "what's with the difference between changeset and all"?

Reply:
I thought I answered that for changeset if I am touching tests I wanted them linted. I can exclude the tests there as well or add test lint as it's own extra flag.

Reply:
ah, sorry.  got to the point where i understand the keyword now!  i'm just concerned that if you add a test in a file that is hard to lint, you'll get all those warnings, but then again you can ignore them.

maybe explain it better in the script?  let me try!

Reply:
```suggestion
              files=$(find libs/ services/ -not -path "*/test/*" -name "*.hs")
              echo "WARNING: not linting tests."
            elif [ "$f" = "changeset" ]; then
              files=$(git diff --name-only HEAD | grep \.hs\$)
              echo "WARNING: linting all test files with changes.  this may lead to some hard to fix warnings/errors, it is safe to ignore these!"
```